### ISSUE TITLE
Proposal: add "use" to anonymous blocks

### ISSUE THREAD

**Issue by lukepighetti** [dart-lang/sdk#35499]
Hello,

I have been using anonymous blocks to break up long procedural code that will not be reused anywhere. I ran into an interesting idea for a language feature that I think would be an interesting addition to Dart. I can imagine there being a lot of apathy towards this proposal due to procedural programming being seen as "old", but since all applications use procedural to some extent I think its worth considering.

Here is a function using anonymous blocks. Its not a shining example, but it gets us started! The nice thing about anonymous blocks is that they limit the scope within steps in procedural code while preserving the order to the reader. Do this, then this, then this. All the code is there any easy to reason about. The problem is that these anonymous blocks perform shadowing and its not clear what variables from the parent scope are being used.

```dart
myFunction(){
  final foo = "foo";
  
  String bar;

  {
    final result = "$foo bar";
    bar = result;
  }

  String baz;

  {
    final result = "$bar baz";
    baz = result;
  }

  return baz;
}
```

Many people would rewrite this as below. Problem is, its not sequential, harder to understand (although so common many would argue this is not a consideration), and makes non-repetitive code very repetitive in a way that normally supports reuse but in this case reuse is not valued.

```dart
_generateBar(String foo){
  final result = "$foo bar";
  return result;
};

_generateBaz(String bar){
  final result = "$bar baz";
  return result;
};

myFunction(){
  final foo = "foo";

  final bar = _generateBar(String foo);
  final baz = _generateBaz(String baz);

  return baz;
}
```

**Proposal: Allow strict scope for anonymous blocks.**

Introduce a `use` keyword that restricts the scope of an anonymous block to the variables provided to `use`.

```dart
myFunction(){
  final foo = "foo";

  final bar = use foo {
    final result = "$foo bar";
    return result;
  }

  final bar = use bar {
    final result = "$bar baz";
    return result;
  }

  return baz;
}
```

An example with multiple arguments might be:

```
final a = "a";
final b = "b";

final c = use a, b {
  return "$a $b";
}
```

**Comment by srawlins** [comment]
CC @eernstg and @leafpetersen I think this could be moved to the language repo.

I actually just like even a simpler concept: let blocks (anonymous, try, ...) return values. This helps assign final and non-nullable variables:

```dart
final a = {
  if (foo.complicatedExpression || otherThings.something) {
    return 1;
  } else if (some.other.conditions) {
    return 2;
  } else {
    return 3;
  }
}

String /* ! */ b = try {
  // something about a connection;
} on NetworkException catch (e) {
  // guaranteed block exit like throw.
  throw MyDifferentException;
  // or a special value
  return "MAGIC_STRING_VALUE";
}
```

**Comment by lukepighetti** [comment]
Although, you can already achieve this like so:

```dart

myFunction(){
  final foo = "foo";

  final bar = (){
    final result = "$foo bar";
    return result;
  };

  final baz = (){
    final result = "$bar baz";
    return result;
  }();

  return baz;
}

```

**Comment by miyoyo** [comment]
@srawlins Your first example is doable today with closures
```dart
final a = (){
  if (foo.complicatedExpression || otherThings.something) {
    return 1;
  } else if (some.other.conditions) {
    return 2;
  } else {
    return 3;
  }
}();
```

```dart
String /* ! */ b = (){
  try {
    // something about a connection;
  } on NetworkException catch (e) {
    // guaranteed block exit like throw.
    throw MyDifferentException;
    // or a special value
    return "MAGIC_STRING_VALUE";
  }
}();
```

The last one is also slightly ambiguous, as there are two blocks
```dart
try {
  // something about a connection;
}
```

and

```dart
on NetworkException catch (e) {
  // guaranteed block exit like throw.
  throw MyDifferentException;
  // or a special value
  return "MAGIC_STRING_VALUE";
}
```

This can be confusing syntax. Also, this would make this use case more complex to understand, and more prone to syntax errors
```dart
String test() {
  try {
    // something about a connection;
  } on NetworkException catch (e) { 
    return "MAGIC_STRING_VALUE";
  }
}
```
```dart
String test() {
  var b = try {
    // something about a connection;
  } on NetworkException catch (e) { 
    return "MAGIC_STRING_VALUE";
  }
}
```
These two are very similar and the error can be glanced over

**Comment by lukepighetti** [comment]
Yes the original idea was to restrict the ability for a block to access outside variables regardless of them being local or global. This would just promote writing blocks as pure-ish functions.

I really like the idea you are proposing. I wanted to add another (orthogonal?) idea for consideration.

It seems to me that in dart there is only one block modifier `async` such as `foo() async { /* */ }`

What if we were to add another, perhaps `pure`

Example:

```dart
foo(String input) pure => "$input bar"; // is valid
foo(String input) pure => "$input, $global, bar"; // is invalid
```

pure would force a function to only consume its inputs

**Comment by eernstg** [comment]
Drive-by comment: We have had several requests for constant function literals, [all the way back to 2012](https://github.com/dart-lang/sdk/issues/4596), and they would be required to denote the same function also in the case where more than one entity is associated with a name from an enclosing scope. This would imply that local variables and instance variables from enclosing scopes cannot be accessed, but it would allow access to static and top-level variables, and it wouldn't prevent side-effects.

@lukepighetti, would that address the request of this issue?
